pypeit/edgetrace.py

```diff
@@ -0,0 +1,4109 @@
+                image; see :func:`initial_trace` for how this used.
+            binning (:obj:`str`, optional):
+                On-detector binning of the data ordered spectral then
+                spatial with format, e.g., `2,1`. Ignored if `img` is
+                an instance of :class:`pypeit.traceimage.TraceImage`.
+            save (:obj:`bool`, optional):
+                Save the result to the master frame.
+            debug (:obj:`bool`, optional):
+                Run in debug mode.
+            show_stages (:obj:`bool`, optional):
+                After ever stage of the auto trace, execute
+                :func:`show` to show the results. These calls to show
+                are always::
+
+                    self.show(thin=10, include_img=True, idlabel=True)
+
+        """
+        # Perform the initial edge detection and trace identification
+        self.initial_trace(img, bpm=bpm, det=det, binning=binning, save=False)
+        if show_stages:
+            self.show(thin=10, include_img=True, idlabel=True)
+
+        # Initial trace can result in no edges found
+        if not self.is_empty:
+            # Refine the locations of the trace using centroids of the
+            # features in the Sobel-filtered image.
+            self.centroid_refine()
+            if show_stages:
+                self.show(thin=10, include_img=True, idlabel=True)
+
+        # Initial trace can result in no edges found, or centroid
+        # refinement could have removed all traces (via `check_traces`)
+        if not self.is_empty:
+            # Fit the trace locations with a polynomial
+            self.fit_refine(debug=debug)
+            # Use the fits to determine if there are any discontinous
+            # trace centroid measurements that are actually components
+            # of the same slit edge
+            self.merge_traces(debug=debug)
+            if show_stages:
+                self.show(thin=10, include_img=True, idlabel=True)
+
+        # Check if the PCA decomposition is possible; this should catch
+        # long slits
+        if self.can_pca():
+            # Use a PCA decomposition to parameterize the trace
+            # functional forms
+            self.pca_refine(debug=debug)
+            if show_stages:
+                self.show(thin=10, include_img=True, idlabel=True)
+
+            # Use the results of the PCA decomposition to recify and
+            # detect peaks/troughs in the spectrally collapsed
+            # Sobel-filtered image, then use those peaks to further
+            # refine the edge traces
+            self.peak_refine(rebuild_pca=True, debug=debug)
+            if show_stages:
+                self.show(thin=10, include_img=True, idlabel=True)
+
+        elif self.par['sync_predict'] == 'pca':
+            # TODO: Maybe we should instead force a fault here instead
+            # of letting the code change a parameter?
+            msgs.warn('Sync predict cannot use PCA; using nearest instead.')
```

Comment:
Okay will do.  Already ran into a problem this caused in `test_scripts.py` where I have to edit the this parameter algorithmically in the pypeit file (harder than I thought it would be) before running the `run_pypeit` tests.